src/main/kotlin/pl/edu/agh/gem/internal/job/stage/GenerateReportStage.kt

```diff
@@ -0,0 +1,45 @@
+package pl.edu.agh.gem.internal.job.stage
+
+import org.bson.types.Binary
+import pl.edu.agh.gem.internal.job.ProcessingStage
+import pl.edu.agh.gem.internal.job.ReportJobState.FETCHING_GROUP_DETAILS
+import pl.edu.agh.gem.internal.job.ReportJobState.FETCHING_SETTLEMENTS
+import pl.edu.agh.gem.internal.job.ReportJobState.FETCHING_USERS_DETAILS
+import pl.edu.agh.gem.internal.job.ReportJobState.SAVING
+import pl.edu.agh.gem.internal.job.StageResult
+import pl.edu.agh.gem.internal.model.finance.GroupActivities
+import pl.edu.agh.gem.internal.model.finance.GroupBalances
+import pl.edu.agh.gem.internal.model.finance.GroupSettlements
+import pl.edu.agh.gem.internal.model.group.GroupDetails
+import pl.edu.agh.gem.internal.model.report.ReportJob
+import pl.edu.agh.gem.internal.model.user.UsersDetails
+
+abstract class GenerateReportStage : ProcessingStage() {
+    abstract fun generateReport(
+        balances: List<GroupBalances>,
+        settlements: List<GroupSettlements>,
+        activities: List<GroupActivities>,
+        usersDetails: UsersDetails,
+        groupDetails: GroupDetails,
+    ): Binary
+
+    override fun process(reportJob: ReportJob): StageResult {
+        return if (reportJob.balances == null || reportJob.settlements == null || reportJob.activities == null) {
```

Comment:
Like in other stage, we don’t have 100% certainty here.We processing it asynchronous, soo It would be better to check and retry from the beginning rather than failing the entire job, which would prevent the user report from being sent.